File: packages/core/src/marginalia_core/ports/capabilities.py
"""Shared provider capability models."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class ProviderExecutionMode(str, Enum):
    """How a provider is expected to execute."""

    LOCAL = "local"
    HYBRID = "hybrid"
    REMOTE = "remote"


@dataclass(frozen=True, slots=True)
class ProviderCapabilities:
    """Capability flags used by fake and future real providers."""

    provider_name: str
    interface_kind: str
    supported_languages: tuple[str, ...] = ("en",)
    supports_streaming: bool = False
    supports_partial_results: bool = False
    supports_timestamps: bool = False
    low_latency_suitable: bool = False
    offline_capable: bool = True
    execution_mode: ProviderExecutionMode = ProviderExecutionMode.LOCAL

File: crates/marginalia-core/src/ports/capabilities.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderExecutionMode {
    Local,
    Hybrid,
    Remote,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProviderCapabilities {
    pub provider_name: String,
    pub interface_kind: String,
    pub supported_languages: Vec<String>,
    pub supports_streaming: bool,
    pub supports_partial_results: bool,
    pub supports_timestamps: bool,
    pub low_latency_suitable: bool,
    pub offline_capable: bool,
    pub execution_mode: ProviderExecutionMode,
}

impl Default for ProviderCapabilities {
    fn default() -> Self {
        Self {
            provider_name: String::new(),
            interface_kind: String::new(),
            supported_languages: vec!["en".to_string()],
            supports_streaming: false,
            supports_partial_results: false,
            supports_timestamps: false,
            low_latency_suitable: false,
            offline_capable: true,
            execution_mode: ProviderExecutionMode::Local,
        }
    }
}
